adds missing formats in the registry (#3167)
OAI/OpenAPI-Specification: registries/_format/binary.md

description: any sequence of octets
base_type: string
layout: default
oas_version: < 3.1

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead set the `mediaType` to `application/octet-stream`, `image/png`,... without a `schema` property. This format MUST ONLY be used for schemas that represent the whole request/response body and MUST NOT be used for properties, arrays, etc..."

Reviewer: Streamlining:

> In OpenAPI 3.1, instead set the media type appropriately and do not use a `schema` property.  Note that only complete HTTP message bodies or complete parts in a `multipart` media type can accommodate binary data.  JSON strings and URL path components cannot.

You can link part of the text to the part of the spec that shows the example.

I'll defer to the @OAI/tsc , but I'd avoid capital MUST in the registry as it is not a normative document.

Author: updated

base_type: string
layout: default
oas_version: < 3.1

> remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead set the `mediaType` to `application/octet-stream`, `image/png`,... without a `schema` property. This format MUST ONLY be used for schemas that represent the whole request/response body and MUST NOT be used for properties, arrays, etc..."

Reviewer 2: ```suggestion
remarks: "When using OpenAPI 3.1 or above it's recommended not to use this format and instead set the `mediaType` to `application/octet-stream`, `image/png`, ... without a `schema` property. This format should only be used for schemas that represent the whole request/response body and should not be used for properties, arrays, etc."
```

Base type: `{{ page.base_type }}`.

> The `{{page.slug}}` format represents any sequence of octets. This format entry is to ensure future versions of OpenAPI maintain compatibility with [OpenAPI 3.0.x](https://spec.openapis.org/oas/v3.0.0).

Reviewer: I think it would be worth noting that in OAS 3.1 this is better written as `"contentMediaType": "application/octet-stream"` without a `"type"` (because binary data is outside of the JSON Schema data model).

Author: right, except when you have scenarios when API return a mix of structured and unstructured data and they are not using a multipart response/body.
A good example of that is a scenario of search result, where both the structured payload and a thumbnail image are returned in a single call.
So should we add a not to suggest people using a different media type is better when the full payload is binary?

Reviewer: I'm not sure I follow.  What is the media type of the response in your example?  You cannot embed raw binary in `application/json`.

Reviewer: @baywet this comment is attached to the `"format": "binary"` definition but your example uses `"format": "base64url"` so now I am even more confused.

Author: Thanks, this clarifies confusion on my end, kiota is loosing precision on that front by mapping binary, base64 and base64url to a mushed "binary" format which it shouldn't be doing since the serialization/deserialization is going to be different (one is straight stream and invalid for properties, others are stream + serialization logic).

We'll have to retro-fit that aspect.

The example provided above applies to base64 and base64url only then.
I'll update this one shortly to provide additional guidance. Let me know what you think.

Author: https://github.com/OAI/OpenAPI-Specification/pull/3167/commits/f45235b8c9c8860c95197eeb8e700af1f2072a0e

Base type: `{{ page.base_type }}`.

The `{{page.slug}}` format represents any sequence of octets. This format entry is to ensure future versions of OpenAPI maintain compatibility with [OpenAPI 3.0.x](https://spec.openapis.org/oas/v3.0.0).

### Upgrading from < 3.1 to >= 3.1

#### Entire response body

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema:
                type: string
                format: binary

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:

>               schema: true #optional, some linters require a schema property

Reviewer: Great idea to show how to work with linters here!

```

#### Encoded property as part of a structured response

> Using the binary format as part of a structured response for a property or an array is invalid. The only exception is for `multipart/form-data` content.

Reviewer: Is "structured response" common usage to mean JSON?  Does it typically include things like `multipart/form-data`?  I would not think of `application/json` and `multipart/form-data` as the same sort of thing, but I spend too much time using RFCs.

Author: JSON or XML or YAML, etc... anything that can be used as a serialization format more or less.

Reviewer: (I'd mark this resolved but apparently I can't)

Author: resolved

The `{{page.slug}}` format represents any sequence of octets. This format entry is to ensure future versions of OpenAPI maintain compatibility with [OpenAPI 3.0.x](https://spec.openapis.org/oas/v3.0.0).

### Upgrading from < 3.1 to >= 3.1

#### Entire response body

```yaml
#OAS 3.0
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema:
                type: string
                format: binary

#OAS 3.1
paths:
  '/picture':
    get:
      responses:
        '200':
          content:
            image/png:
              schema: true #optional, some linters require a schema property
```

#### Encoded property as part of a structured response

> Using the binary format as part of a structured response for a property or an array is invalid. The only exception is for `multipart/form-data` content.

Author: ```suggestion
Using the binary format as part of a structured response for a property or an array is invalid. Use `multipart/form-data` content instead.
```

#### Query or Path parameters

> Using the binary format query or path parameters is invalid.

Reviewer 2: ```suggestion
Using the binary format for query or path parameters is invalid.
```
